# Student Performance Indicator

## Life cycle of Machine learning Project

1. Understanding the Problem Statement
2. Data Collection
3. Data Checks to perform
4. Exploratory data analysis
5. Data Pre-Processing
6. Model Training
7. Choose best model

## 1) Problem statement

This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

## 2) Data Collection

Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
The data consists of 8 column and 1000 rows.

## 2.1 Import Data and Required Packages

* Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [1]:
import numpy as np
import pandas as pd 
import seaborn as sns 
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

* Import the CSV Data as Pandas DataFrame

In [2]:
df = pd.read_csv('data/raw.csv')

* Showing Top 5 records

In [3]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

* Shape of the dataset


In [4]:
df.shape

(1002, 8)

# 2.2 Dataset Information

* gender : sex of students -> (Male/female)
* race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
* parental level of education : parents' final education ->(bachelor's degree some college,master's degree,associate's degree,high school)
* lunch : having lunch before test (standard or free/reduced)
* test preparation course : complete or not complete before test
* math score
* reading score
* writing score

# 3. Data Checks to perform

* Check Missing values
* Check Duplicates
* Check data type
* Check the number of unique values of each column
* Check statistics of data set
* Check various categories present in the different categorical column

## 3.1 Checking Missing Values

In [5]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

* There are no missing values in the dataset 

## 3.2 Checking Duplicates

In [6]:
df.duplicated().sum()


np.int64(1)

* There are no duplicate values in the dataset

## 3.3 Checking data types

In [7]:
# checking Null and Dtypes

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1002 entries, 0 to 1001
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1002 non-null   object
 1   race_ethnicity               1002 non-null   object
 2   parental_level_of_education  1002 non-null   object
 3   lunch                        1002 non-null   object
 4   test_preparation_course      1002 non-null   object
 5   math_score                   1002 non-null   int64 
 6   reading_score                1002 non-null   int64 
 7   writing_score                1002 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.8+ KB


## 3.4 Checking the number of unique values of each column

In [8]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

## 3.5 Checking statistics of dataset

In [9]:
df.describe()

math_score  reading_score  writing_score
count  1002.000000    1002.000000    1002.000000
mean     66.111776      69.202595      68.089820
std      15.156509      14.604950      15.201606
min       0.000000      17.000000      10.000000
25%      57.000000      59.000000      58.000000
50%      66.000000      70.000000      69.000000
75%      77.000000      79.750000      79.000000
max     100.000000     100.000000     100.000000

### Insight

* From above description of numerical data, all means are very close to each other - between 66 and 68.05;
* All standard deviations are also close - between 14.6 and 15.19;
* While there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17

## 3.7 Exploring Data

In [10]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [11]:
print("Categories in 'gender' variable:  ",end=" " )
print(df['gender'].unique())

print("Categories in 'race_ethnicity' variable:  ",end=" ")
print(df['race_ethnicity'].unique())

print("Categories in'parental level of education' variable:  ",end=" " )
print(df['parental_level_of_education'].unique())

print("Categories in 'lunch' variable:  ",end=" " )
print(df['lunch'].unique())

print("Categories in 'test preparation course' variable:  ",end=" " )
print(df['test_preparation_course'].unique())

Categories in 'gender' variable:   ['female' 'male']
Categories in 'race_ethnicity' variable:   ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in'parental level of education' variable:   ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable:   ['standard' 'free/reduced']
Categories in 'test preparation course' variable:   ['none' 'completed']


In [12]:
# Defining numerical and categorical features

numeric_features=[feature for feature in df.columns if df[feature].dtype !='O']
categorical_features=[feature for feature in df.columns if df[feature].dtype == 'O']

# print columns

print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 3 numerical features : ['math_score', 'reading_score', 'writing_score']

We have 5 categorical features : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


## 3.8 Adding columns for "Total Score" and "Average"

In [13]:
df['total score']=df['math_score']+df['reading_score']+df['writing_score']
df['average']=df['total score']/3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

### Checking how many got full marks

In [14]:
reading_full_marks = df[df['reading_score'] == 100].shape[0]
writing_full_marks = df[df['writing_score'] == 100].shape[0]
math_full_marks = df[df['math_score'] == 100].shape[0]
print(f"Number of students who scored full marks in Reading: {reading_full_marks}")
print(f"Number of students who scored full marks in Writing: {writing_full_marks}")
print(f"Number of students who scored full marks in Math: {math_full_marks}")

Number of students who scored full marks in Reading: 17
Number of students who scored full marks in Writing: 14
Number of students who scored full marks in Math: 7


### Checking how many got less than 20 marks

In [15]:
reading_less_20= df[df['reading_score'] <= 20].shape[0]
writing_less_20= df[df['writing_score'] <= 20].shape[0]
math_less_20= df[df['math_score'] <= 20].shape[0]
print(f"Number of students who scored less than 20 in Reading: {reading_less_20}")
print(f"Number of students who scored less than 20 in Writing: {writing_less_20}")
print(f"Number of students who scored less than 20 in Math: {math_less_20}")

Number of students who scored less than 20 in Reading: 1
Number of students who scored less than 20 in Writing: 3
Number of students who scored less than 20 in Math: 4


#### Insights

* From above values we get students have performed the worst in Maths
* Best performance is in reading section

## 4. Exploring Data ( Visualization )

### 4.1 Visualising average score distribution to make some conclusion.

* Histogram
* Kernel Distribution Function (KDE)

#### 4.1.1 Histogram & KDE